ports/servoshell/desktop/webview.rs

```diff
@@ -149,18 +121,41 @@ where
         self.webviews.get_mut(&webview_id)
     }
 
-    // Returns the existing preload data for the given WebView, or a new one.
-    fn ensure_preload_data_mut(&mut self, webview_id: &WebViewId) -> &mut WebViewPreloadData {
-        if let Entry::Vacant(entry) = self.webview_preload_data.entry(*webview_id) {
-            entry.insert(WebViewPreloadData::default());
-        }
-        self.webview_preload_data.get_mut(webview_id).unwrap()
+    pub fn get(&self, webview_id: WebViewId) -> Option<&WebView> {
+        self.webviews.get(&webview_id)
+    }
+
+    pub(crate) fn add(&mut self, webview: ::servo::WebView) {
+        self.creation_order.push(webview.id());
+        self.webviews.insert(webview.id(), WebView::new(webview));
     }
 
     pub fn focused_webview_id(&self) -> Option<WebViewId> {
         self.focused_webview_id
     }
 
+    pub fn close_webview(&mut self, servo: &Servo, webview_id: WebViewId) {
+        // This can happen because we can trigger a close with a UI action and then get the
+        // close event from Servo later.
+        if !self.webviews.contains_key(&webview_id) {
+            return;
+        }
+
+        self.webviews.retain(|&id, _| id != webview_id);
```

Comment:
this will yield warnings about closing a webview twice, when a webview closes of its own volition. that’s ok, we’re moving towards a model where only the embedder can close webviews soon anyway.